fl4health/losses/mkmmd_loss.py

```diff
@@ -77,156 +80,218 @@ def construct_quadruples(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor
         )
-        y_y_prime[y_y_prime < 0] = 0
         x_y_prime = (
             torch.sum((X**2), dim=1).reshape(-1, 1)
             + torch.sum((Y**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(X, torch.transpose(Y, 0, 1))
         )
-        x_y_prime[x_y_prime < 0] = 0
         x_prime_y = (
             torch.sum((Y**2), dim=1).reshape(-1, 1)
             + torch.sum((X**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(Y, torch.transpose(X, 0, 1))
         )
+
+        # Correct any values that ended up nearly but not identically zero (||x-y||^2 should be semi-definite)
+        x_x_prime[x_x_prime < 0] = 0
+        y_y_prime[y_y_prime < 0] = 0
+        x_y_prime[x_y_prime < 0] = 0
         x_prime_y[x_prime_y < 0] = 0
-        # each inner product is a tensor of dimension len(v_i_quadruples) x len(v_i_quadruples), we return a
+
+        # each inner product is a tensor of dimension n_samples x n_samples, we return a
         # tensor of shape 4 x len(X) x len(Y)
         return torch.cat(
-            [x_x_prime.unsqueeze(0), y_y_prime.unsqueeze(0), x_y_prime.unsqueeze(0), x_prime_y.unsqueeze(0)], dim=0
+            [x_x_prime.unsqueeze(0), y_y_prime.unsqueeze(0), x_y_prime.unsqueeze(0), x_prime_y.unsqueeze(0)]
         )
 
+    def compute_euclidean_inner_products_linear(self, v_i_quadruples: torch.Tensor) -> torch.Tensor:
+        # Shape of v_i_quadruples is n_samples/2 x 4 x n_features
+        # v_i = [x_{2i-1}, x_{2i}, y_{2i-1}, y_{2i}]
+        #
+        # We want to compute the RBF kernel values. To do this, we need to compute ||x - y||^2 for the relevant pairs
+        # x and y. That is the inner product. Note that ||x - y||^2 = <x - y, x - y> = (x-y)^T(x-y)
+        # For the quadruples of the form (x,  x', y, y') we need distances for pairings (x, x'), (y, y'), (x, y'),
+        # (x, y')
+        x_x_prime = torch.sum((v_i_quadruples[:, 0, :] - v_i_quadruples[:, 1, :]) ** 2, dim=1, keepdim=True)
+        y_y_prime = torch.sum((v_i_quadruples[:, 2, :] - v_i_quadruples[:, 3, :]) ** 2, dim=1, keepdim=True)
+        x_y_prime = torch.sum((v_i_quadruples[:, 0, :] - v_i_quadruples[:, 3, :]) ** 2, dim=1, keepdim=True)
+        x_prime_y = torch.sum((v_i_quadruples[:, 1, :] - v_i_quadruples[:, 2, :]) ** 2, dim=1, keepdim=True)
+
+        # each inner product is a tensor of dimension len(v_i_quadruples), we return a tensor of shape
+        # len(v_i_quadruples) x 4
+        return torch.cat([x_x_prime, y_y_prime, x_y_prime, x_prime_y], dim=1)
+
     def compute_h_u_from_inner_products(self, inner_products: torch.Tensor, gamma: torch.Tensor) -> torch.Tensor:
         # Gamma should be of shape torch.Tensor([gamma])
         assert gamma.shape == (1,)
-        # inner_products has shape number of 4 x v_i_quadruples x v_i_quadruples or 4 x len(X) x len(Y)
-        # h_u_componets should have the same shape
+        # inner_products has shape of 4 x n_samples x n_samples
```

Comment:
Technically `compute_h_u_from_inner_products` and `compute_h_u_from_inner_products_linear` have the same computations but the underlying values being acted on have different structure. So I broke them apart to allow for distinct comments.